File: js/src/0.0-Basics/0.3-build-up-logical-thinking/0.3.20-pattern.js
function log(val) {
    process.stdout.write(String(val));
}

function pattern_20() {
    let chars = ['a', 'b', 'c', 'd', 'e'];

    let row = 5;
    let column = 10;

    for (let y = 0; y <= row - 1; y++) {
        for (let x = column - 1; x > 0; x--) {
            if (x > column / 2) {
                let xIndex = x - column / 2;

                if (xIndex <= y) {
                    log(chars[xIndex]);
                } else {
                    log(' ');
                }
            } else if (x <= column / 2) {
                let yIndex = column / 2 - x;

                if (yIndex <= y) {
                    log(chars[yIndex]);
                } else {
                    log(' ');
                }
            }
        }
        console.log();
    }
}

pattern_20();

File: python/00-Basics/00.03-build-up-logical-thinking/0.3.20-pattern.py
from utils.clear import clear

def pattern_20():
    row = 5
    column = 10

    chars = ['a', 'b', 'c', 'd', 'e']

    for i in range(0, row ):
        for j in range(1, column):
            if j <= column / 2:
                key = round(column / 2 - j) 
                if key <= i:
                    print(chars[key], end="")
                elif key > i:
                    print(" ", end="")

            if j > column / 2:
                key = round(j - column / 2)
                if key <= i:
                    print(chars[key], end="")
                elif key > i:
                    print(" ", end="")

        print()


clear()
pattern_20()
